python/packages/core/agent_framework/_workflows/_handoff.py

```diff
@@ -0,0 +1,1388 @@
+
+
+def _as_user_messages(payload: Any) -> list[ChatMessage]:
+    if isinstance(payload, ChatMessage):
+        if payload.role == Role.USER:
+            return [payload]
+        return [ChatMessage(Role.USER, text=payload.text)]
+    if isinstance(payload, list):
+        # Check if all items are ChatMessage instances
+        all_chat_messages = all(isinstance(msg, ChatMessage) for msg in payload)  # type: ignore[arg-type]
+        if all_chat_messages:
+            messages: list[ChatMessage] = payload  # type: ignore[assignment]
+            return [msg if msg.role == Role.USER else ChatMessage(Role.USER, text=msg.text) for msg in messages]
+    if isinstance(payload, Mapping):  # User supplied structured data
+        text = payload.get("text") or payload.get("content")  # type: ignore[union-attr]
+        if isinstance(text, str) and text.strip():
+            return [ChatMessage(Role.USER, text=text.strip())]
+    return [ChatMessage(Role.USER, text=str(payload))]  # type: ignore[arg-type]
+
+
+def _default_termination_condition(conversation: list[ChatMessage]) -> bool:
+    """Default termination: stop after 10 user messages to prevent infinite loops."""
+    user_message_count = sum(1 for msg in conversation if msg.role == Role.USER)
+    return user_message_count >= 10
+
+
+class HandoffBuilder:
+    r"""Fluent builder for conversational handoff workflows with triage and specialist agents.
+
+    The handoff pattern models a customer support or multi-agent conversation where:
+    - A **triage/dispatcher agent** receives user input and decides whether to handle it directly
+      or hand off to a **specialist agent**.
+    - After a specialist responds, control returns to the user for more input, creating a cyclical flow:
+      user -> triage -> [optional specialist] -> user -> triage -> ...
+    - The workflow automatically requests user input after each agent response, maintaining conversation continuity.
+    - A **termination condition** determines when the workflow should stop requesting input and complete.
+
+    Routing Patterns:
+
+    **Single-Tier (Default):** Only the triage agent can hand off to specialists. After any specialist
+    responds, control returns to the user. This is the recommended pattern for most scenarios because:
+
+    - **User stays in the loop**: Provides visibility and control after each specialist interaction
+    - **Simpler mental model**: Star topology with triage as hub, specialists as spokes
+    - **Prevents scope creep**: Specialists focus on domain expertise, not routing logic
+    - **Easier debugging**: All routing decisions flow through one central agent
+    - **Better error handling**: User can intervene immediately if a specialist fails
+
+    **Multi-Tier (Advanced):** Specialists can hand off to other specialists using `.with_handoffs()`.
+    Use this pattern only when you have genuine business requirements for specialist collaboration:
```

Comment:
Same here, "genuine business requirements" is too ambiguous. I would skip the bullet points below and simply say it's less controllable comparing to the dispatcher-only pattern, but provide more flexibility.

Reply:
Valid, as well. Cleaned this up.